### PR TITLE
Remove panic on corruped storage

### PR DESCRIPTION
**Pull request by thiolliere**
Fix corrupted storage panic

In the long term we should implement https://github.com/paritytech/substrate/issues/3700

### CODE REVIEW

File: srml/support/src/storage/generator/linked_map.rs [hunk 1]
```diff
@@ -151,22 +159,32 @@ where
...
-					it always points to existing keys; qed");
-		res.1.next = linkage.next;
-		unhashed::put(prev_key.as_ref(), &res);
+		if let Some(mut res) = read_with_linkage::<_, _, G>(prev_key.as_ref()) {
+			res.1.next = linkage.next;
+			unhashed::put(prev_key.as_ref(), &res);
+		} else {
+			// TODO #3700: error should be handleable.
+			runtime_print::print(
+				"ERROR: Corrupted state: linked map {}: previous value doesn't exist at {:x}",
```
Comment (bkchr) [inline srml/support/src/storage/generator/linked_map.rs]: We need to call into the else branch below to write the correct head.

Comment (thiolliere) [reply srml/support/src/storage/generator/linked_map.rs]: Yes, I was mixed about what was the best and conclude the least we touch things the best it is.

Here either:
* we let the head where it is, knowing that the list has been cut in two, with second part having a wrong head.
* move the head to next values potentially cutting a large part of the list.

Comment (bkchr) [reply srml/support/src/storage/generator/linked_map.rs]: But this means we will print out this message over and over again. (Given there is no modifcation in between)

Comment (thiolliere) [reply srml/support/src/storage/generator/linked_map.rs]: I don't think this code should attempt to fix stuff, but ok maybe we can do it.

Then we should also maybe:
* remove entry when in storage::get decode has failing, 
* set the next to None when enumeration is broken,

Comment (thiolliere) [reply srml/support/src/storage/generator/linked_map.rs]: oh ya ok at least we can make previous being None for the next element. indeed

**Comment by bkchr** [comment]
Any update?

**Comment by thiolliere** [comment]
actually I'm confused about what to do:
for remove_linkage it seems we could do:
* if both previous key and next key are corrupted (like the linkage what the actual value corrupted that somehow decoded correctly) then do no touch anything
* else if previous key is corrupted then change head and set the previous of the next one to None.
* else if next key is corrupted then set None to the next of the previous one.

Is that correct ? I still kind of prefer not touching anything if possible but if you agree on this then I make it.

**Comment by gavofyork** [comment]
looks strictly better than before, which is enough for a merge.

**Comment by thiolliere** [comment]
Resolved

**Pull request merged**